### PR TITLE
Query API review changes

### LINKED ISSUE #12084: Review internal service dependencies for 3.0
For cases where there should be a dependency object but isn't, and for cases where there are internal/public issues. See #10233 for example.

### LINKED ISSUE #15107: Review individual members documented as internal
Some may be able to be made public now.

### LINKED ISSUE #15695: Final review of internal code usage
Once query is ready, we need to do a final review of internal code used by providers. Everything other than query and conventions should be in a good state, but we may as well check everything one last time when doing the work for query.

### LINKED ISSUE #15796: QueryRewrite: re-enable ApiConsistency tests

### PR DESCRIPTION
**Pull request by ajcvickers**
Mostly moving types to different namespaces (internal or public), and a few renames.

Also, checked where Cosmos and other providers were using internal types and made those pubic.

Re-enabled API consistency, except for NotNull attributes (since these aren't breaking to add later.)

Fixes #12084
Fixes #15107
Fixes #15695
Fixes #15796

### CODE REVIEW

File: src/EFCore/Extensions/EntityTypeExtensions.cs [hunk 1]
```diff
@@ -22,6 +22,9 @@ namespace Microsoft.EntityFrameworkCore
     /// </summary>
     public static class EntityTypeExtensions
     {
+        public static IEnumerable<IEntityType> GetConcreteDerivedTypesInclusive([NotNull] this IEntityType entityType)
```
Comment (smitpatel) [inline src/EFCore/Extensions/EntityTypeExtensions.cs]: Docs

File: src/EFCore/Infrastructure/ExpressionExtensions.cs [hunk 2]
```diff
@@ -23,6 +25,88 @@ namespace Microsoft.EntityFrameworkCore.Infrastructure
     /// </summary>
     public static class ExpressionExtensions
     {
+        private static readonly string _efTypeName = typeof(EF).FullName;
+
+        public static bool IsEFPropertyMethod([CanBeNull] this MethodInfo methodInfo)
```
Comment (smitpatel) [inline src/EFCore/Infrastructure/ExpressionExtensions.cs]: MethodInfo extension on expression extensions?

File: src/EFCore/Metadata/PropertyIdentity.cs [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+using System.Diagnostics;
+using System.Reflection;
+using JetBrains.Annotations;
+using Microsoft.EntityFrameworkCore.Metadata.Internal;
+
+namespace Microsoft.EntityFrameworkCore.Metadata
+{
+    [DebuggerDisplay("{DebuggerDisplay(),nq}")]
+    public readonly struct MemberIdentity
```
Comment (smitpatel) [inline src/EFCore/Metadata/PropertyIdentity.cs]: Filename

File: src/EFCore/Infrastructure/ExpressionExtensions.cs (hunk 2 above)
Comment (ajcvickers) [reply src/EFCore/Infrastructure/ExpressionExtensions.cs]: Will move.

File: src/EFCore/Metadata/PropertyIdentity.cs (hunk 3 above)
Comment (ajcvickers) [reply src/EFCore/Metadata/PropertyIdentity.cs]: Will fix.

File: src/EFCore/Query/ExpressionPrinter.cs [hunk 4]
```diff
@@ -51,57 +42,21 @@ public class ExpressionPrinter : ExpressionVisitor
...
-        /// </summary>
         public virtual IndentedStringBuilder StringBuilder => _stringBuilder;
 
-        /// <summary>
-        ///     This is an internal API that supports the Entity Framework Core infrastructure and not subject to
-        ///     the same compatibility standards as public APIs. It may be changed or removed without notice in
-        ///     any release. You should only use it directly in your code with extreme caution and knowing that
-        ///     doing so can result in application failures when updating to a new Entity Framework Core release.
-        /// </summary>
         public virtual bool RemoveFormatting { get; set; }
```
Comment (smitpatel) [inline src/EFCore/Query/ExpressionPrinter.cs]: Make anything not used outside of this class private.

Comment (ajcvickers) [reply src/EFCore/Query/ExpressionPrinter.cs]: Will do.

**Pull request merged**